climada/util/lines_polys_handler.py

```diff
@@ -0,0 +1,1059 @@
+    -------
+    gdf_pnt : gpd.GeoDataFrame
+        with a double index, first for the geometries of exp, second for the point disaggregation
+        of the geometries.
+
+    """
+    if gdf.empty:
+        return gdf
+
+    pnt_mask, line_mask, poly_mask = _pnt_line_poly_mask(gdf)
+
+    # Concatenating an empty dataframe with an index together with
+    # a dataframe with a multi-index breaks the multi-index
+    gdf_pnt = gpd.GeoDataFrame([])
+    if pnt_mask.any():
+        gdf_pnt = gpd.GeoDataFrame(pd.concat([
+            gdf_pnt,
+            gdf[pnt_mask]
+        ]))
+    if line_mask.any():
+        gdf_pnt = gpd.GeoDataFrame(pd.concat([
+            gdf_pnt,
+            _line_to_pnts(gdf[line_mask], res, to_meters)
+        ]))
+    if poly_mask.any():
+        gdf_pnt = gpd.GeoDataFrame(pd.concat([
+            gdf_pnt,
+            _poly_to_pnts(gdf[poly_mask], res, to_meters)
+        ]))
+
+    return gdf_pnt
+
+
+def gdf_to_grid(gdf, grid):
+    """
+    Disaggregate geodataframe with (multi-)polygons geometries to points.
+
+    Parameters
+    ----------
+    gdf : gpd.GeoDataFrame
+        Feodataframe instance with gdf.geometry containing (multi)-lines or (multi-)polygons.
+        Points are ignored.
+    grid : np.array()
+        Grid on which to disaggregate the exposures. Provided as two vectors [x_grid, y_grid].
+
+    Returns
+    -------
+    gdf_pnt : gpd.GeoDataFrame
+        with a double index, first for the geometries of exp, second for the point disaggregation
+        of the geometries.
+    """
+    if gdf.empty:
+        return gdf
+
+    pnt_mask, line_mask, poly_mask = _pnt_line_poly_mask(gdf)
+
+    # Concatenating an empty dataframe with an index together with
+    # a dataframe with a multi-index breaks the multi-index
+    gdf_pnt = gpd.GeoDataFrame([])
+    if pnt_mask.any():
+        gdf_pnt = gpd.GeoDataFrame(pd.concat([
+            gdf_pnt,
+            gdf[pnt_mask]
+        ]))
+    if line_mask.any():
+        raise AttributeError("The dataframe contains lines. Lines cannot be disaggregated onto a "
+                             "fixed grid.")
+    if poly_mask.any():
+        gdf_pnt = gpd.GeoDataFrame(pd.concat([
+            gdf_pnt,
+            _poly_to_grid(gdf[poly_mask], grid)
+        ]))
+
+    return gdf_pnt
+
+
+def _disagg_values_div(gdf_pnts):
+    """
+    Disaggregate value column of original gdf to disaggregated point gdf by
+    dividing value from geometry equally on points.
+
+    Parameters
+    ----------
+    gdf_pnts : gpd.GeoDataFrame
+        Geodataframe with a double index, first for geometries (lines, polygons),
+        second for the point disaggregation of the polygons. The value column is assumed
+        to represent values per polygon / line (first index).
+
+    Returns
+    -------
+    gdf_disagg : gpd.GeoDataFrame
+        The value per geometry are evenly distributed over the points per geometry.
+
+    """
+
+    gdf_disagg = gdf_pnts.copy()
```

Comment:
Is it really necessary to make a deep copy here? I doubt `groupy` makes any changes on the original dataframe and `reindex` is making a copy of the index by itself (unless called with `copy=False`)

Reply:
sounds reasonable.